src/trace/trace.c

```diff
@@ -96,18 +98,120 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	platform_shared_commit(trace, sizeof(*trace));
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	/* LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE high */
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+struct recent_log_entry {
+	uint32_t entry_id;
+	uint64_t message_ts;
+	uint64_t first_suppression_ts;
+	uint32_t suppress_count;
+};
+
+static struct recent_log_entry recent_entries[CONFIG_TRACE_RECENT_ENTRIES_COUNT];
+static struct recent_log_entry entry_to_emit;
```

Comment:
Actually, I think that a better solution would be to put recent entries in core context, and track them on each core separately. That wouldn't require any locks, and different cores wouldn't interfere with their filters.
What do you think about such solution?

Reply:
Where do we do the ordering and merging of trace messages (based on timestamp) from across cores ?

Reply:
IIUC, right now each core would keep recent entries cached and work individually for as long as cache stays valid.
@lgirdwood  As I've mentioned before, we could create per-core structures to decouple cores from each other. That should be a rather quick change.
In general, we don't expect to see repeated messages, that would have to be suppressed, that pass the filter if tracked on each core separately, but would be suppressed if combined from many cores together.

Reply:
if you do this per-core, wouldn't you get inconsistent logs? E.g. if the same message prints on two cores from some global service, e.g. from a DMA driver, one core already started suppressing that message, another hasn't yet. As well as other similar disagreements.

Reply:
@lyakh @lgirdwood doing this per-core has few advantages, like ensuring that we don't suppress 'golden trace' on one core, because of flood on the other core. Moreover, giving each core control over it's messages' suppression avoids synchronization between cores and simplifies the code, while each core can filter it's flooding messages just fine by itself.